src/librustc/traits/object_safety.rs

```diff
@@ -323,9 +316,108 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
             return Some(MethodViolationCode::WhereClauseReferencesSelf(span));
         }
 
+        let receiver_ty = self.liberate_late_bound_regions(
+            method.def_id,
+            &sig.map_bound(|sig| sig.inputs()[0]),
+        );
+
+        // until we get by-value DST, `self: Self` can't be coerced
+        // but we allow this as a special case.
+        // maybe instead we should also check for
+        // for (U) { if (Self: Unsize<U>) { Receiver: Unsize<Receiver<Self=Self>>}}
+        if receiver_ty != self.mk_self_type() {
+            if !self.receiver_is_coercible(method, receiver_ty) {
+                return Some(MethodViolationCode::UncoercibleReceiver);
+            }
+        }
+
         None
     }
 
+    // checks the type of the self argument, and makes sure it implements
+    // the CoerceUnsized requirement:
+    // forall (U) {
+    //     if (Self: Unsize<U>) {
+    //         Receiver: CoerceUnsized<Receiver<Self=U>>
+    //     }
+    // }
+    #[allow(dead_code)]
+    fn receiver_is_coercible(
+        self,
+        method: &ty::AssociatedItem,
+        receiver_ty: Ty<'tcx>,
+    ) -> bool
```

Comment:
Nit: `{` goes on the line above, no?

Reply:
Good point. Although this whole file is formatted weird, I should really run rustfmt on it at some point.